crates/ruff_python_semantic/src/scope.rs

```diff
@@ -63,18 +67,28 @@ impl<'a> Scope<'a> {
             self.shadowed_bindings.insert(id, shadowed);
             Some(shadowed)
         } else {
+            self.declared_symbols.push(name);
             None
         }
     }
 
-    /// Returns `true` if this scope defines a binding with the given name.
-    pub fn defines(&self, name: &str) -> bool {
+    /// Removes the binding with the given name.
+    pub fn delete(&mut self, name: &'a str) -> Option<BindingId> {
+        self.declared_symbols.push(name);
+        self.bindings.remove(name)
```

Comment:
This actually isn't sufficient, because this needs to be considered valid (for the purpose of the `nonlocal` rule):

```py
def f():
    del x

    def g():
        nonlocal x
```

`del x` adds `x` to the symbol table, but doesn't bind it to the value (or something like that -- I don't have the terminology nailed down).